### PR TITLE
Sections 1-3

### PR DESCRIPTION
**Pull request by jverce**
# Description
This is the first of a series of PR's that document the guidelines
to follow during a code review. It covers the following sections:
1. Preliminary steps
2. Sensible definition of props
3. Good code structure
4. Good 3rd party API usage

# Code Changes
* Add new docs section for PR review guidelines
* Include VuePress config file in the linting process
* Include line numbers to code blocks in the docs
* Update dependencies

### CODE REVIEW

File: docs/docs/components/guidelines/README.md [hunk 1]
```diff
@@ -239,7 +239,7 @@ Registry components are organized by app in the `components` directory of the
   `/components/twitter/sources/search-mentions/search-mentions.js`.
 
 You can explore examples in the [components
-directory](https://github.com/pipedreamhq/pipedream/tree/master/components).
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:242]: Broken link. Do you need a `/code` prefix here, too?

File: docs/docs/components/guidelines/README.md [hunk 2]
```diff
@@ -249,7 +249,27 @@ constructed API requests to reduce code and improve maintenance.
 
 #### Pagination
 
-When making API requests, handle pagination to ensure all data / events are processed.
+When making API requests, handle pagination to ensure all data/events are
+processed. Moreover, if the underlying account experiences and/or generates too
+much data paginating through the entire collection of records might cause
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:254]: paginating through the entire collection of records might -> 

paginating through the entire collection of records, it might

File: docs/docs/components/guidelines/README.md [hunk 3]
```diff
@@ -268,39 +288,51 @@ out](https://pipedream.com/community/c/dev/11).
...
-file.
+Whenever possible, reuse [methods](/components/api/#methods) defined in the app
+file or within a [common module](#common-files-optional).
 
-Use the [JS Docs](https://jsdoc.app/about-getting-started.html) pattern for
-lightweight documentation of each method in the app file. Provide a description
-and define @params and @returns block tags (with default values if applicable —
-e.g., `[foo=bar]`). This data will both help with reusability and will be
-surfaced in documentation for apps in the Pipedream marketplace. For example:
+As a rule of thumb, every interaction with the app-specific API should go
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:303]: Technically we don't call the `.app` files "components", so I would change to "app file" and link to `#app-files`

File: docs/docs/components/guidelines/README.md [hunk 4]
```diff
@@ -268,39 +288,51 @@ out](https://pipedream.com/community/c/dev/11).
...
+through the `.app` component. If you need to use an API for which a method is
+not defined and it may be used in future components, define a new method in the
+`.app` component.
+
+Behavior that is specific to the components themselves (e.g. a method to
+generate the metadata of an emitted event) can be specified in a common module
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:309]: Can you link "common module" to the common module anchor below?

File: docs/docs/components/guidelines/README.md [hunk 5]
```diff
@@ -324,17 +356,42 @@ module.exports = {
...
 
 Pipedream does not currently support unit tests to validate that changes to app
 files are backwards compatible with existing components. Therefore, if you make
 changes to an app file that may impact other sources, you must currently test
 potentially impacted components to confirm their functionality is not negatively
 affected. We expect to support a testing framework in the future.
 
+Some tips on what to test:
+
+- High traffic, or large record collections:
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:369]: Small feedback but can you make the top-level bullets here (the sections) `#####` tags, instead? That'll let us link to them as anchors and I think the section delineating reads easier than nested bullets.

File: docs/docs/components/guidelines/README.md [hunk 6]
```diff
@@ -346,28 +403,181 @@ approach, the general pattern is:
...
+directly reutilize the attributes defined by the common modules as well as
+allowing for any further refinement if necessary.
+
+To illustrate this idea, let's assume that we have an app called **My App**,
+which is already available in Pipedream under the `my_app` key, meaning that all
+of its source code lives under `components/my_app`.
+
+- `components/my_app/my_app.app.js`
+
+  This is the main `.app` module which defines behavior and properties related
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:428]: `.app` module -> app file

File: docs/docs/components/guidelines/README.md [hunk 7]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+few preliminary steps that we'd like you to take in order to avoid
+misunderstandings, wasted time and effort:
+
+1. Targeting a PR:
+   1. Check the [main Kanban board](/code/projects/1) for any tasks pending
+      review (i.e. the **Ready for PR review** column). These tasks should be
+      associated to an open PR that is waiting for a review.
+   2. If the above step does not yield any PR, explicitly look for open pull
+      requests in the [pull requests page](/code/pulls) without reviews and
+      without reviewers assigned to it. [This
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:781]: Maybe remove the final sentence here and make "open pull requests without reviews and without reviewers assigned to it" the link.

File: docs/docs/components/guidelines/README.md [hunk 8]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+on. In this case it makes much more sense to instead create a component named
+something like `cloud_storage-new-file-of-type` that is customizable via a prop
+of type `string[]` named `fileTypes` and let users select the types of files
+they prefer to watch.
+
+The advantage of making a module customizable is that it can then be extended
+into particular use cases very easily should those be highly demanded. See the
+[Code Structure](#code-structure) section for an example.
+
+There is no clear line as to which components should be customizable and which
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:852]: I would rephrase this section slightly to say, "If it's not clear what components you should create, reach out to the Pipedream team on the issue". We actually have a Product Manager for the team now who will be making these calls, and normally we'll have an opinion on it.

File: docs/docs/components/guidelines/README.md [hunk 9]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+ones should be a sole component, so use your best judgment here to check that
+the component's props make sense both from a maintenance effort as well as from
+a user's point of view.
+
+#### Code Structure
+
+An important part of the review involves making sure that the code is properly
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:859]: I would shorten to "Code should be properly structured so that:"

File: docs/docs/components/guidelines/README.md [hunk 10]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+
+::: danger
+This approach is not ideal since it duplicates logic and can cause future
+maintenance pain and confusion
+:::
+
+A better approach would be to extract any common logic outside of the specific
+component (i.e. `cloud_storage-new-file`) so that both this component as well as
+any new ones (e.g. `cloud_storage-new-pdf-document`) become easier to maintain
+as they become smaller while also allowing for future similar event sources to
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:935]: Add a comma between "smaller" and "while"

File: docs/docs/components/guidelines/README.md [hunk 11]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+      isRelevant({ type }) {
+        return this.fileTypes.includes(type);
+      },
+    },
+  };
+  ```
+
+- `cloud_storage/sources/new-pdf-document/new-pdf-document.js`
+
+  This component is a simplified version of the above, which also extends the
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:1034]: Add a period after "module" and before "but".

Add a comma between "interest" and "it"

File: docs/docs/components/guidelines/README.md [hunk 12]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+      isRelevant() {
+        return true;
+      },
+    },
+  };
+  ```
+
+#### Good API Usage
+
+In the previous sections we highlighted the fact that the `.app` modules are the
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:1083]: `.app` modules -> "app files" and link to #app-files anchor

File: docs/docs/components/guidelines/README.md [hunk 13]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+ones that contain almost all of the logic regarding the interaction with the
+underlying 3rd party app. For example, the
+[`gitlab.app`](/code/components/gitlab/gitlab.app.js) module defines several
+methods that encapsulate calls to and knowledge about the [Gitlab
+API](https://docs.gitlab.com/ee/api/).
+
+Given that the `.app` modules provide such abstractions around external API's,
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:1090]: Can you replace instances of "API's" with "APIs"? We generally use the latter in these docs since it's not a possessive noun.

File: docs/docs/components/guidelines/README.md [hunk 14]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+          yield files;
+        }
+      },
+    },
+  };
+  ```
+
+  ::: tip
+  Note that the interface around this API call only requires the drive ID, and
+  hides the details regarding the pagination details. Any additional parameter
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:1324]: "hides the details regarding the pagination details" -> "hides the details regarding pagination"

File: docs/docs/components/guidelines/README.md [hunk 15]
```diff
@@ -552,4 +762,823 @@ end user. Generate and use a GUID for the shared secret value, save it to a
...
+third party API it **must always offload as much processing as possible** when
+it comes to such operations. Some of the advantages of this approach are:
+
+- Offloading computation efforts to the third party service
+- Computations performed on the service side are likely performed on the entire
+  collection set. If a component were to achieve a similar result it would
+  likely have to retrieve all the data from the service, store it in memory,
+  pre-process it (e.g. sorting the records) and then proceed to the actual
+  record processing.
+- Search API's usually provide certain consistency guarantees which is then
```
Comment (dylburger) [inline docs/docs/components/guidelines/README.md:1403]: "which is" -> "which are" since guarantees is plural

**Review by dylburger — CHANGES_REQUESTED** [review]
@jverce mostly small changes

File: docs/docs/components/guidelines/README.md (hunk 1 above)
Comment (jverce) [reply docs/docs/components/guidelines/README.md:242]: Oops, my bad 😊 
Fixed!

File: docs/docs/components/guidelines/README.md (hunk 7 above)
Comment (jverce) [reply docs/docs/components/guidelines/README.md:783]: You mean remove this sentence?

> [This link](/code/pulls?q=is%3Apr+is%3Aopen+review%3Anone+no%3Aassignee) already provides a filtered view for the mentioned criteria.

File: docs/docs/components/guidelines/README.md (hunk 8 above)
Comment (jverce) [reply docs/docs/components/guidelines/README.md:854]: Rephrasing the whole paragraph to this, WDYT?

> If it's unclear which components should be customizable and which ones should be
a sole component, reach out to the Pipedream team on the issue since this
usually requires input from a product perspective.

**Comment by GTFalcao** [comment]
Closing as outdated: docs is no longer in this repo
